Importing the Dependencies

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [4]:

credit_card_data = pd.read_csv('/content/creditcard.csv')

In [5]:

credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [6]:
credit_card_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
213964  139469.0 -0.313355  1.038943 -0.093972 -0.540988  0.042780 -1.106101   
213965  139469.0  2.030213  0.285221 -1.938915  1.027250  0.851942 -0.701464   
213966  139470.0  0.142229  0.672251  0.983900 -0.293528  0.068882 -0.487888   
213967  139471.0  2.144033 -1.953243 -0.578686 -1.849212 -1.376208  0.807803   
213968  139471.0  2.157143 -0.687990 -2.032580 -0.940456 -0.215889       NaN   

              V7        V8        V9  ...       V21       V22       V23  \
213964  0.603947  0.287733 -0.100080  ... -0.217842 -0.621741  0.153701   
213965  0.615953 -0.241818 -0.349579  ...  0.167846  0.510669  0.027440   
213966  0.491696 -0.118390  0.318579  ...  0.000967  0.224127  0.213429   
213967 -1.911412  0.306645 -0.726623  ... -0.124630 -0.092680  0.193302   
213968       NaN       NaN       NaN  ...       NaN       NaN       NaN   

             V24       V25       V26       V27       V28  Amount  Class  
213964 -0.042585 -0.503706  0.148254  0.118168  0.032668    9.12    0.0  
213965  0.689138  0.427077 -0.532637 -0.040637 -0.068185    2.90    0.0  
213966 -0.107864 -1.475182 -0.749164  0.330808  0.289856    8.57    0.0  
213967 -0.404247 -0.484276 -0.247000  0.034838 -0.033122   84.00    0.0  
213968       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [7]:

credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 213969 entries, 0 to 213968
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    213969 non-null  float64
 1   V1      213969 non-null  float64
 2   V2      213969 non-null  float64
 3   V3      213969 non-null  float64
 4   V4      213969 non-null  float64
 5   V5      213969 non-null  float64
 6   V6      213968 non-null  float64
 7   V7      213968 non-null  float64
 8   V8      213968 non-null  float64
 9   V9      213968 non-null  float64
 10  V10     213968 non-null  float64
 11  V11     213968 non-null  float64
 12  V12     213968 non-null  float64
 13  V13     213968 non-null  float64
 14  V14     213968 non-null  float64
 15  V15     213968 non-null  float64
 16  V16     213968 non-null  float64
 17  V17     213968 non-null  float64
 18  V18     213968 non-null  float64
 19  V19     213968 non-null  float64
 20  V20     213968 non-null  float64
 21  V21     21

In [8]:

credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        1
V7        1
V8        1
V9        1
V10       1
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [9]:

credit_card_data['Class'].value_counts()

Class
0.0    213570
1.0       398
Name: count, dtype: int64

This Dataset is highly unblanced

0 --> Normal Transaction

1 --> fraudulent transaction

In [10]:

legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [11]:
print(legit.shape)
print(fraud.shape)

(213570, 31)
(398, 31)


In [12]:

legit.Amount.describe()

count    213570.000000
mean         90.156783
std         248.673400
min           0.000000
25%           6.000000
50%          23.165000
75%          79.640000
max       19656.530000
Name: Amount, dtype: float64

In [13]:
fraud.Amount.describe()

count     398.000000
mean      123.841307
std       257.528158
min         0.000000
25%         1.000000
50%        13.385000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [14]:

credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    74958.173451 -0.071181 -0.013179  0.261446  0.049393 -0.076991   
1.0    63734.716080 -5.539063  4.099702 -7.613034  4.720286 -3.954560   

             V6       V7        V8        V9  ...       V20       V21  \
Class                                         ...                       
0.0    0.041862 -0.02447  0.008461  0.010928  ...  0.013739 -0.011748   
1.0   -1.383136 -6.45837  0.665504 -2.735173  ...  0.391298  0.776397   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.034001 -0.012882  0.002689  0.048307  0.003970  0.000165  0.001790   
1.0   -0.002832 -0.045213 -0.071156  0.061446  0.041531  0.185812  0.064843   

           Amount  
Class              
0.0     90.156783  
1.0    123.841307  

[2 rows x 30 columns]

Under-Sampling

Build a sample dataset containing similar distribution of normal transactions and Fraudulent Transactions

Number of Fraudulent Transactions --> 492

In [15]:
legit_sample = legit.sample(n=492)

Concatenating two DataFrames

In [16]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [ ]:
new_dataset.head()

Time        V1        V2  ...       V28  Amount  Class
203131  134666.0 -1.220220 -1.729458  ... -0.023852  155.00      0
95383    65279.0 -1.295124  0.157326  ...  0.105345   70.00      0
99706    67246.0 -1.481168  1.226490  ...  0.076538   40.14      0
153895  100541.0 -0.181013  1.395877  ... -0.329608  137.04      0
249976  154664.0  0.475977 -0.573662  ...  0.012816   19.60      0

[5 rows x 31 columns]

In [17]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
208651  137211.0  0.630579  1.183631 -5.066283  2.179903 -0.703376 -0.103614   
212516  138894.0 -1.298443  1.948100 -4.509947  1.305805 -0.019486 -0.509238   
212644  138942.0 -2.356348  1.746360 -6.374624  1.772205 -3.439294  1.457811   
213092  139107.0 -4.666500 -3.952320  0.206094  5.153525  5.229469  0.939040   
213116  139117.0 -3.975939 -1.244939 -3.707414  4.544772  4.050676 -3.407679   

              V7        V8        V9  ...       V21       V22       V23  \
208651 -3.490350  1.094734 -0.717418  ...  0.621622  0.043807  0.102711   
212516 -2.643398  1.283545 -2.515356  ...  1.178032  1.360989 -0.272013   
212644 -0.362577  1.443791 -1.927359  ...  0.857942  0.621203  0.964817   
213092 -0.635033 -0.704506 -0.234786  ... -0.664263  1.821422  0.113563   
213116 -5.063118  1.007042 -3.190158  ...  1.059737 -0.037395  0.348707   

             V24       V25       V26       V27       V28  Amount  Class  
208651 -0.601505  0.127371 -0.163009  0.853792  0.356503   39.45    1.0  
212516 -0.325948  0.290703  0.841295  0.643094  0.201156    0.01    1.0  
212644 -0.619437 -1.732613  0.108361  1.130828  0.415703  727.91    1.0  
213092 -0.759673 -0.502304  0.630639 -0.513880  0.729526   22.47    1.0  
213116 -0.162929  0.410531 -0.123612  0.877424  0.667568    8.30    1.0  

[5 rows x 31 columns]

In [18]:
new_dataset['Class'].value_counts()

Class
0.0    492
1.0    398
Name: count, dtype: int64

In [19]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4       V5  \
Class                                                                  
0.0    75637.756098 -0.026998 -0.087131  0.175222  0.106843 -0.07571   
1.0    63734.716080 -5.539063  4.099702 -7.613034  4.720286 -3.95456   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.077451 -0.038711  0.023872 -0.035128  ...  0.009631 -0.008680   
1.0   -1.383136 -6.458370  0.665504 -2.735173  ...  0.391298  0.776397   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.018458  0.008784 -0.025171  0.027620 -0.017525 -0.053990  0.025504   
1.0   -0.002832 -0.045213 -0.071156  0.061446  0.041531  0.185812  0.064843   

           Amount  
Class              
0.0     94.646016  
1.0    123.841307  

[2 rows x 30 columns]

Splitting the data into Features & Targets

In [20]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [21]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
92344    63889.0 -1.444318 -0.482830  1.245921 -1.547637 -1.673127 -0.693946   
33642    37317.0 -0.466743  0.835188  0.683234  0.046954 -0.336993 -0.395224   
99926    67340.0 -1.189569  0.613020  1.777271 -0.387916  0.571143  2.884494   
188872  128170.0  2.138155  0.624783 -3.600220  0.363175  1.800702 -0.745528   
163309  115842.0  1.818376 -0.802293 -0.562986  0.177397 -0.702708 -0.144587   
...          ...       ...       ...       ...       ...       ...       ...   
208651  137211.0  0.630579  1.183631 -5.066283  2.179903 -0.703376 -0.103614   
212516  138894.0 -1.298443  1.948100 -4.509947  1.305805 -0.019486 -0.509238   
212644  138942.0 -2.356348  1.746360 -6.374624  1.772205 -3.439294  1.457811   
213092  139107.0 -4.666500 -3.952320  0.206094  5.153525  5.229469  0.939040   
213116  139117.0 -3.975939 -1.244939 -3.707414  4.544772  4.050676 -3.407679   

              V7        V8        V9  .

In [22]:
print(Y)

92344     0.0
33642     0.0
99926     0.0
188872    0.0
163309    0.0
         ... 
208651    1.0
212516    1.0
212644    1.0
213092    1.0
213116    1.0
Name: Class, Length: 890, dtype: float64


Split the data into Training data & Testing Data

In [23]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [24]:
print(X.shape, X_train.shape, X_test.shape)

(890, 30) (712, 30) (178, 30)


Model Training

Logistic Regression

In [ ]:
model = LogisticRegression()

In [ ]:

model.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Model Evaluation

Accuracy Score

In [ ]:

X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [ ]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9415501905972046


In [ ]:

X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [ ]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.9390862944162437


In [26]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

mo = RandomForestClassifier(n_estimators=50, max_depth=10, min_samples_split=5, min_samples_leaf=2, random_state=42)
mo.fit(X_train, Y_train)
X_train_prediction = mo.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print("Accuracy on Training Data:", training_data_accuracy)

X_test_prediction = mo.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print("Accuracy on Test Data:", test_data_accuracy)


Accuracy on Training Data: 0.9817415730337079
Accuracy on Test Data: 0.9269662921348315


In [28]:
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score


mo = XGBClassifier(n_estimators=50, max_depth=3, learning_rate=0.05, subsample=0.6, colsample_bytree=0.8, random_state=42)
mo.fit(X_train, Y_train)

X_train_prediction = mo.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print("Accuracy on Training Data:", training_data_accuracy)

X_test_prediction = mo.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print("Accuracy on Test Data:", test_data_accuracy)


Accuracy on Training Data: 0.9648876404494382
Accuracy on Test Data: 0.9269662921348315


In [29]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

mo = SVC(kernel='rbf', C=1.0, gamma='scale', random_state=42)
mo.fit(X_train, Y_train)

X_train_prediction = mo.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print("Accuracy on Training Data:", training_data_accuracy)

X_test_prediction = mo.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print("Accuracy on Test Data:", test_data_accuracy)


Accuracy on Training Data: 0.6067415730337079
Accuracy on Test Data: 0.5730337078651685


#                    Accuracy using different models


**Logistic Regression**

                  Accuracy on Training data :0.9415501905972046
                  Accuracy score on Test Data :0.939086294162437

**Random Forest**

                  Accuracy on Training Data: 0.9817415730337079
                  Accuracy on Test Data: 0.9269662921348315


  **eXtreme Gradient Boosting**
                 
                 Accuracy on Training Data: 0.9648876404494382
                 Accuracy on Test Data: 0.9269662921348315


  **Support Vector Classifier**

                 Accuracy on Training Data: 0.6067415730337079
                 Accuracy on Test Data: 0.5730337078651685


#      Logistic Regression is best with 93.90% accuracy               

